python/paddle/fluid/dygraph/io.py

```diff
@@ -742,91 +744,100 @@ def _construct(model_path, configs=None):
+                            % var_name)
+
+                output_vars = []
+                for var_desc in program_holder.output_descs:
+                    var = core.VarBase(var_desc.dtype(),
+                                       var_desc.shape(),
+                                       var_desc.name(), var_desc.type(), False)
+                    output_vars.append(var)
+
+                # hold forward variables
+                tmp_scope_vec = core.VarBase(
+                    core.VarDesc.VarType.FP32, [], "program_out_scope",
+                    core.VarDesc.VarType.STEP_SCOPES, True)
+                tmp_scope_vec.value().set_scope(program_holder.scope)
+
+                # 2. run program by op
+                trace_program = program_holder.infer_program if self._is_test else program_holder.train_program
+                end_op_index = program_holder.infer_program.block(0).op_size()
+                framework._dygraph_tracer().trace_op(
+                    type='run_program',
+                    inputs={'X': input_vars,
+                            'Params': persistable_vars},
+                    outputs={'Out': output_vars,
+                             'OutScope': tmp_scope_vec},
+                    attrs={
+                        'global_block': trace_program.block(0),
+                        'start_op_index': 0,
+                        'end_op_index': end_op_index,
+                        'is_test': self._is_test
+                    })
+                # NOTE: [ why need set param's gradient type here ]
+                # if user set sparse gradient mode, the param's gradient
+                # will be SelectedRows, not LoDTensor. But tracer will just
+                # set param grad VarBase by forward VarBase(LoDTensor)
+                # If we don't change grad_var type here, RunProgramOp need
+                # transform SelectedRows to LoDTensor forcibly, it may not
+                # be user wanted result.
+                for persistable_var in persistable_vars:
+                    grad_var_name = var.name + core.grad_var_suffix()
+                    grad_var = trace_program.block(0).find_var(
+                        cpt.to_bytes(grad_var_name))
+                    # NOTE: cannot find var desc maybe not problem, 
+                    # such as in batch_norm
+                    if grad_var is None:
+                        continue
+                    persistable_var._set_grad_type(grad_var.type())
+
+                # 3. prepare output, keep same form with inputs
+                outs = output_vars
+                if len(output_vars) == 1:
+                    outs = output_vars[0]
+                return outs
+            else:
+                trace_program = program_holder.infer_program if self._is_test else program_holder.train_program
+                p = framework.Program._construct_from_desc(trace_program)
```

Comment:
why need construct program here, why not use `trace_program` directly

Reply:
Done,thx.

Reply:
Done,thx.

Reply:
'_run_static_graph' modifies the 'trace_program'.  If use 'trace_program' directly, a segmentation error will occur.